File: ios/UsageLimitsKit/Sources/UsageLimitsKit/Network/ProviderEndpoints.swift
import Foundation

/// Every provider URL, client identifier and request header the app sends, in one place.
///
/// Ported from the Android `ProviderEndpoints.kt`, value for value. These are reverse-engineered
/// from the first-party CLI and desktop clients by way of CLIProxyAPI and its Management Center
/// (see `docs/provider-auth-research.md` for the exact source files and commits). They are
/// **not** a documented third-party API and can change without notice, which is why they are
/// centralised: an upstream change should be a one-file edit on each platform, never a hunt
/// through parsers.
///
/// Nothing here is a secret. The OAuth client IDs are the public identifiers the official
/// clients ship; see `Antigravity.clientSecret` for the one value that needs a caveat.
public enum ProviderEndpoints {

    /// OpenAI Codex / ChatGPT subscription.
    public enum Codex {
        public static let clientID = "app_EMoamEEZ73f0CkXaXp7hrann"

        /// Authorization code + PKCE with a loopback redirect — the flow the Codex CLI itself
        /// runs, and the default on both platforms. The client registration pins this exact
        /// redirect, so the app listens on this port just as it does for Claude and
        /// Antigravity. The extra parameters are the ones the first-party client sends; the
        /// authorize page shapes what it shows by them.
        public static let authorizeURL = "https://auth.openai.com/oauth/authorize"
        public static let redirectURI = "http://localhost:1455/auth/callback"
        public static let authorizeScope = "openid profile email offline_access"
        public static let authorizeExtraParameters = [
            "id_token_add_organizations": "true",
            "codex_cli_simplified_flow": "true",
            "originator": "codex_cli_rs",
        ]

        // Device authorization — the fallback when the redirect port is taken. No redirect URI
        // has to be registered and no loopback listener has to run. The provider generates the
        // PKCE pair and returns it with the code, which means PKCE here is not the
        // client-binding guarantee it normally is; that caveat is recorded in the research doc.
        public static let deviceUserCodeURL = "https://auth.openai.com/api/accounts/deviceauth/usercode"
        public static let deviceTokenURL = "https://auth.openai.com/api/accounts/deviceauth/token"
        public static let deviceVerificationURL = "https://auth.openai.com/codex/device"

        /// Redirect the device-issued code must be exchanged against. Never navigated to.
        public static let deviceExchangeRedirectURI = "https://auth.openai.com/deviceauth/callback"

        public static let tokenURL = "https://auth.openai.com/oauth/token"

        public static let usageURL = "https://chatgpt.com/backend-api/wham/usage"
        public static let resetCreditsURL = "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits"
        public static let resetCreditsConsumeURL = "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits/consume"

        /// Identifies the app as a Codex client. The backend varies its response by client, so
        /// this is a compatibility marker rather than an attempt to look like something else.
        public static let userAgent = "codex-tui/0.149.1 (iOS; arm64) UsageLimits"

        /// Sent on the refresh grant only. Neither the device flow nor the code exchange uses it.
        public static let refreshScope = "openid profile email"

        /// Selects the account on the usage and reset-credit calls. Omitted entirely when the
        /// real `chatgpt_account_id` is unknown — a JWT `sub` is a user id, not an account id,
        /// and sending it scopes the response to the wrong account.
        public static let accountIDHeader = "Chatgpt-Account-Id"

        /// Only the reset-credit endpoints take these; the usage endpoint does not.
        public static let resetCreditHeaders = [
            "OpenAI-Beta": "codex-1",
            "Originator": "Codex Desktop",
        ]
    }

    /// Claude / Anthropic subscription.
    public enum Claude {
        public static let clientID = "9d1c250a-e61b-44d9-88ed-5944d1962f5e"
        public static let authorizeURL = "https://claude.ai/oauth/authorize"

        /// Both the code exchange and the refresh go here — `platform.claude.com`, not `api.`.
        public static let tokenURL = "https://platform.claude.com/v1/oauth/token"

        public static let profileURL = "https://api.anthropic.com/api/oauth/profile"
        public static let usageURL = "https://api.anthropic.com/api/oauth/usage"

        /// On iOS this is a custom scheme rather than the loopback URI Android must use:
        /// `ASWebAuthenticationSession` handles the callback in-process, which removes the
        /// pinned-port design and every failure that came with it.
        ///
        /// It must still be registered with the provider to be accepted; until it is, the
        /// loopback URI remains the only value that works, and that is a real open question
        /// for the iOS port rather than a settled detail.
        public static let redirectURI = "http://localhost:54545/callback"
        public static let callbackScheme = "usagelimits"

        public static let scope =
            "user:profile user:inference user:sessions:claude_code user:mcp_servers user:file_upload"

        public static let betaHeader = "oauth-2025-04-20"

        /// Usage window keys, in display order.
        ///
        /// `iguana_necktie` is upstream's current key for the Fable weekly window — a codename,
        /// not a typo.
        public static let usageWindowKeys: [(key: String, label: String)] = [
            ("five_hour", "5h limit"),
            ("seven_day", "Weekly"),
            ("seven_day_oauth_apps", "Weekly (OAuth apps)"),
            ("seven_day_opus", "Weekly (Opus)"),
            ("seven_day_sonnet", "Weekly (Sonnet)"),
            ("seven_day_cowork", "Weekly (Cowork)"),
            ("iguana_necktie", "Weekly (Fable)"),
        ]
    }

    /// Google Antigravity.
    public enum Antigravity {
        public static let clientID =
            "1071006060591-tmhssin2h21lcre235vtolojh4g403ep.apps.googleusercontent.com"

        /// Google "installed application" client secret.
        ///
        /// Not a confidential credential: RFC 8252 §8.5 and Google's own installed-app
        /// documentation treat it as public, and it ships in the desktop client already. It is
        /// required because Google's token endpoint rejects the exchange without it for this
        /// client type, and the Antigravity quota scopes are bound to this specific client — a
        /// self-registered client cannot reach them. PKCE is what actually protects the flow.
        /// See `docs/security.md` for the full argument and the rejected alternatives.
        public static let clientSecret = "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"

        public static let authEndpoint = "https://accounts.google.com/o/oauth2/v2/auth"
        public static let tokenEndpoint = "https://oauth2.googleapis.com/token"
        public static let userInfoEndpoint = "https://www.googleapis.com/oauth2/v2/userinfo?alt=json"

        public static let redirectURI = "http://localhost:51121/oauth-callback"
        public static let callbackScheme = "usagelimits"

        public static let scopes = [
            "https://www.googleapis.com/auth/cloud-platform",
            "https://www.googleapis.com/auth/userinfo.email",
            "https://www.googleapis.com/auth/userinfo.profile",
            "https://www.googleapis.com/auth/cclog",
            "https://www.googleapis.com/auth/experimentsandconfigs",
        ]

        /// Tried in order. The stable host is first because this call gates account creation.
        public static let loadCodeAssistURLs = [
            "https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist",
            "https://daily-cloudcode-pa.googleapis.com/v1internal:loadCodeAssist",
        ]

        /// Tried in order until one answers with data. The daily and sandbox hosts are rolled
        /// out ahead of the stable one, and which host serves a given account varies.
        public static let quotaURLs = [
            "https://daily-cloudcode-pa.googleapis.com/v1internal:retrieveUserQuotaSummary",
            "https://daily-cloudcode-pa.sandbox.googleapis.com/v1internal:retrieveUserQuotaSummary",
            "https://cloudcode-pa.googleapis.com/v1internal:retrieveUserQuotaSummary",
        ]

        public static let cliVersion = "1.0.13"
        public static let userAgent = "antigravity/cli/\(cliVersion) (aidev_client; os_type=darwin; arch=arm64)"

        /// `ideType` selects which product's onboarding record Cloud Code returns.
        /// `IDE_UNSPECIFIED` is the *Gemini Code Assist* value and yields a project carrying
        /// none of the Antigravity quota groups.
        public static let loadCodeAssistMetadata: [String: String] = [
            "ideType": "ANTIGRAVITY",
            "platform": "PLATFORM_UNSPECIFIED",
            "pluginType": "GEMINI",
        ]
    }

    /// xAI / Grok.
    public enum Xai {
        public static let clientID = "b1a00492-073a-47ea-816f-4c329264a828"
        public static let issuer = "https://auth.x.ai"

        /// Endpoints are discovered rather than hardcoded, then validated: HTTPS only, and the
        /// host must be exactly `x.ai` or a dot-anchored `.x.ai` subdomain. The anchor is what
        /// stops `notx.ai` and `x.ai.example.com` passing.
        public static let discoveryURL = "\(issuer)/.well-known/openid-configuration"
        public static let issuerHost = "x.ai"

        public static let scope = "openid profile email offline_access grok-cli:access api:access"
        public static let deviceCodeGrantType = "urn:ietf:params:oauth:grant-type:device_code"

        public static let billingCreditsURL = "https://cli-chat-proxy.grok.com/v1/billing?format=credits"
        public static let billingURL = "https://cli-chat-proxy.grok.com/v1/billing"
        public static let meURL = "https://api.x.ai/v1/me"

        public static let clientVersion = "0.2.91"
        public static let userAgent = "grok-pager/\(clientVersion) grok-shell/\(clientVersion) (ios; aarch64)"

        public static let identityHeaders = [
            "x-xai-token-auth": "xai-grok-cli",
            "x-grok-client-version": clientVersion,
        ]
    }

    /// Kimi Code.
    ///
    /// OAuth, under this app's OWN name. Kimi's device flow uses one public client id for every
    /// program that drives it, and what `api.kimi.com/coding` gates on is the `X-Msh-Platform`
    /// header naming the calling program. Moonshot allowlists third-party programs on request
    /// and forbids exactly one thing: presenting another program's identity. So this app says
    /// who it is, never `kimi_cli`. Until its name is allowlisted the coding API may answer
    /// `403 access_terminated`, and a key from the user's own console stays the other way in.
    /// See docs/providers-kimi.md.
    public enum Kimi {
        /// The public client id every Kimi Code device-flow client presents.
        public static let clientID = "17e5f671-d194-4dfb-9706-5516cb48c098"
        public static let deviceCodeURL = "https://auth.kimi.com/api/oauth/device_authorization"
        public static let tokenURL = "https://auth.kimi.com/api/oauth/token"
        public static let deviceCodeGrantType = "urn:ietf:params:oauth:grant-type:device_code"

        public static let usageEndpoint = "https://api.kimi.com/coding/v1/usages"
        public static let consoleURL = "https://www.kimi.com/code"

        /// This app's own name, sent wherever Kimi asks which program is calling.
        public static let platform = "UsageLimits"

        /// Who is calling, truthfully. The version is the app's own, read from its bundle.
        public static var identityHeaders: [String: String] {
            let version = (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String)
                .flatMap { $0.isEmpty ? nil : $0 } ?? "0"
            return [
                "User-Agent": "\(platform)/\(version) (iOS)",
                "X-Msh-Platform": platform,
                "X-Msh-Version": version,
            ]
        }
    }

    /// Devin / Cognition.
    ///
    /// Devin's CLI uses a PKCE authorization-code flow with a loopback callback. The callback
    /// port is part of the URL the browser signs, so the iOS adapter uses a dedicated local port
    /// and includes it in every authorization request. The quota endpoint is a Connect-RPC
    /// method on the Codeium server; the client accepts both its protobuf and JSON encodings.
    public enum Devin {
        public static let appBaseURL = "https://app.devin.ai"
        public static let apiBaseURL = "https://api.devin.ai"
        public static let serverBaseURL = "https://server.codeium.com"
        public static let authorizationURL = "\(appBaseURL)/auth/cli/continue"
        public static let tokenURL = "\(apiBaseURL)/auth/cli/token"
        public static let profileURL = "\(apiBaseURL)/v3/self"
        public static let userStatusURL =
            "\(serverBaseURL)/exa.seat_management_pb.SeatManagementService/GetUserStatus"

        /// Devin accepts a caller-selected loopback port. A fixed port keeps the existing iOS
        /// listener API synchronous while remaining local-only; callers can override it when a
        /// conflict is detected by passing a custom endpoint in tests or a future UI flow.
        public static let callbackPort: UInt16 = 19876
        public static let redirectURI = "http://127.0.0.1:\(callbackPort)/callback"

        public static let userAgent = "UsageLimits/1.0 (iOS)"
        public static let ideName = "chisel"
        public static let ideVersion = "3000.10.21"
        public static let locale = "en"
        public static let sessionTokenPrefix = "devin-session-token$"
        public static let fingerprintHexLength = 732
    }

    /// Meta Muse.
    ///
    /// Meta exposes the Muse CLI sign-in as an RFC 8628 device grant. The access token returned
    /// by the device token endpoint is a DCA token; the `muse-code/key` call exchanges it for the
    /// LLM key and returns the subscription/quota metadata used by the app.
    public enum Meta {
        public static let clientID = "1031625952748946"
        public static let authHost = "https://auth.meta.com"
        public static let deviceAuthorizationURL = "\(authHost)/oidc/device/authorization/"
        public static let deviceTokenURL = "\(authHost)/oidc/device/token/"
        public static let deviceVerificationURL = "\(authHost)/device"
        public static let deviceCodeGrantType =
            "urn:ietf:params:oauth:grant-type:device_code"

        public static let keyURL = "https://api.meta.ai/muse-code/key"
        public static let apiBaseURL = "https://api.meta.ai/v1"
        public static let userAgent = "muse-code/1.0.2"
        public static let apiVersion = "1.0.0"
    }
}

File: app/src/main/kotlin/com/usagelimits/core/network/ProviderEndpoints.kt
package com.usagelimits.core.network

/**
 * Every provider URL, client identifier and request header the app sends, in one place.
 *
 * These values are reverse-engineered from the first-party CLI/desktop clients by way of
 * CLIProxyAPI and its Management Center (see docs/provider-auth-research.md for the exact
 * source files and the commits they were read at). They are *not* a documented third-party
 * API and can change without notice, so they are centralised here: an upstream change should
 * be a one-file edit, never a hunt through parsers.
 *
 * Nothing here is a secret. The OAuth client IDs are the public identifiers the official
 * clients ship; see [Antigravity.CLIENT_SECRET] for the one value that needs a caveat.
 */
object ProviderEndpoints {

    /**
     * OpenAI Codex / ChatGPT subscription.
     *
     * Source: CLIProxyAPI internal/auth/codex/openai_auth.go, sdk/auth/codex_device.go, and
     * Management Center src/utils/quota/constants.ts.
     */
    object Codex {
        const val CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"

        /**
         * Authorization code + PKCE with a loopback redirect — the flow the Codex CLI itself
         * runs, and the default here. The client registration pins this exact redirect, so
         * the app binds this port (see LoopbackServer) just as it does for Claude and
         * Antigravity. The extra parameters are the ones the first-party client sends; the
         * authorize page shapes what it shows by them.
         */
        const val AUTHORIZE_URL = "https://auth.openai.com/oauth/authorize"
        const val REDIRECT_PORT = 1455
        const val REDIRECT_URI = "http://localhost:$REDIRECT_PORT/auth/callback"
        const val AUTHORIZE_SCOPE = "openid profile email offline_access"
        val AUTHORIZE_EXTRA_PARAMS = mapOf(
            "id_token_add_organizations" to "true",
            "codex_cli_simplified_flow" to "true",
            "originator" to "codex_cli_rs",
        )

        // Device authorization — the fallback when the redirect port is taken. No redirect
        // URI has to be registered and no loopback server has to run. The provider generates
        // the PKCE pair and returns it with the code.
        const val DEVICE_USER_CODE_URL = "https://auth.openai.com/api/accounts/deviceauth/usercode"
        const val DEVICE_TOKEN_URL = "https://auth.openai.com/api/accounts/deviceauth/token"
        const val DEVICE_VERIFICATION_URL = "https://auth.openai.com/codex/device"

        /** Redirect the device-issued code must be exchanged against. Never navigated to. */
        const val DEVICE_EXCHANGE_REDIRECT_URI = "https://auth.openai.com/deviceauth/callback"

        const val TOKEN_URL = "https://auth.openai.com/oauth/token"

        const val USAGE_URL = "https://chatgpt.com/backend-api/wham/usage"
        const val RESET_CREDITS_URL =
            "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits"
        const val RESET_CREDITS_CONSUME_URL =
            "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits/consume"

        /**
         * Identifies the app as a Codex client. The backend varies its response by client, so
         * this is a compatibility marker rather than an attempt to look like something else.
         */
        const val USER_AGENT = "codex-tui/0.149.1 (Android; arm64) UsageLimits"

        /** Sent on the refresh grant only. Neither the device flow nor the code exchange uses it. */
        const val SCOPE = "openid email profile offline_access"

        /** Sent on the usage and reset-credit calls to select the account. */
        const val HEADER_ACCOUNT_ID = "Chatgpt-Account-Id"

        /**
         * Only the reset-credit endpoints require these; the usage endpoint does not send
         * them. Verified in Management Center codex/data.ts (fetchCodexResetCredits).
         */
        val RESET_CREDIT_HEADERS = mapOf(
            "OpenAI-Beta" to "codex-1",
            "Originator" to "Codex Desktop",
        )
    }

    /**
     * Claude / Anthropic subscription.
     *
     * Source: CLIProxyAPI internal/auth/claude/anthropic_auth.go and Management Center
     * src/utils/quota/constants.ts.
     */
    object Claude {
        const val CLIENT_ID = "9d1c250a-e61b-44d9-88ed-5944d1962f5e"
        const val AUTHORIZE_URL = "https://claude.ai/oauth/authorize"

        /** Both the code exchange and the refresh go here (platform.claude.com, not api.). */
        const val TOKEN_URL = "https://platform.claude.com/v1/oauth/token"

        const val PROFILE_URL = "https://api.anthropic.com/api/oauth/profile"
        const val USAGE_URL = "https://api.anthropic.com/api/oauth/usage"

        /**
         * Loopback redirect, per RFC 8252 for native apps. The port is fixed because the
         * provider's client registration pins it, so the app must bind exactly this port.
         */
        const val REDIRECT_URI = "http://localhost:54545/callback"
        const val REDIRECT_PORT = 54545

        const val SCOPE =
            "user:profile user:inference user:sessions:claude_code user:mcp_servers user:file_upload"

        const val BETA_HEADER = "oauth-2025-04-20"

        /**
         * Usage window keys, in display order.
         *
         * `iguana_necktie` is upstream's current key for the Fable weekly window — a codename,
         * not a typo. The parser tolerates its absence and ignores unknown siblings, so a
         * rename upstream degrades to "one window missing" rather than a failed parse.
         */
        val USAGE_WINDOW_KEYS = listOf(
            "five_hour" to "5h limit",
            "seven_day" to "Weekly",
            "seven_day_oauth_apps" to "Weekly (OAuth apps)",
            "seven_day_opus" to "Weekly (Opus)",
            "seven_day_sonnet" to "Weekly (Sonnet)",
            "seven_day_cowork" to "Weekly (Cowork)",
            "iguana_necktie" to "Weekly (Fable)",
        )
    }

    /**
     * Google Antigravity.
     *
     * Source: CLIProxyAPI internal/auth/antigravity/constants.go and Management Center
     * src/utils/quota/constants.ts.
     */
    object Antigravity {
        const val CLIENT_ID =
            "1071006060591-tmhssin2h21lcre235vtolojh4g403ep.apps.googleusercontent.com"

        /**
         * Google "installed application" client secret.
         *
         * This is not a confidential credential: RFC 8252 §8.5 and Google's own installed-app
         * documentation treat it as public, and it ships in the desktop client already. It is
         * required because Google's token endpoint rejects the exchange without it for this
         * client type, and the Antigravity quota scopes are bound to this specific client —
         * a self-registered Android client cannot reach them.
         *
         * PKCE is still used, so possession of this string alone does not let an attacker
         * complete a flow. See docs/security.md for the full argument and the alternatives
         * that were rejected.
         */
        const val CLIENT_SECRET = "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"

        const val AUTH_ENDPOINT = "https://accounts.google.com/o/oauth2/v2/auth"
        const val TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token"
        const val USERINFO_ENDPOINT = "https://www.googleapis.com/oauth2/v2/userinfo?alt=json"

        const val REDIRECT_PORT = 51121
        const val REDIRECT_URI = "http://localhost:$REDIRECT_PORT/oauth-callback"

        val SCOPES = listOf(
            "https://www.googleapis.com/auth/cloud-platform",
            "https://www.googleapis.com/auth/userinfo.email",
            "https://www.googleapis.com/auth/userinfo.profile",
            "https://www.googleapis.com/auth/cclog",
            "https://www.googleapis.com/auth/experimentsandconfigs",
        )

        /**
         * Tried in order. The stable host is first because this call gates account creation:
         * CLIProxyAPI and the Quota Inspector both use `cloudcode-pa` for it, and only the
         * quota read needs the daily hosts' wider rollout.
         */
        val LOAD_CODE_ASSIST_URLS = listOf(
            "https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist",
            "https://daily-cloudcode-pa.googleapis.com/v1internal:loadCodeAssist",
        )

        /**
         * Tried in order until one answers. The daily/sandbox hosts are rolled out ahead of
         * the stable one, and which host serves a given account varies.
         */
        val QUOTA_URLS = listOf(
            "https://daily-cloudcode-pa.googleapis.com/v1internal:retrieveUserQuotaSummary",
            "https://daily-cloudcode-pa.sandbox.googleapis.com/v1internal:retrieveUserQuotaSummary",
            "https://cloudcode-pa.googleapis.com/v1internal:retrieveUserQuotaSummary",
        )

        const val CLI_VERSION = "1.0.13"
        const val USER_AGENT = "antigravity/cli/$CLI_VERSION (aidev_client; os_type=linux; arch=arm64)"
    }

    /**
     * xAI / Grok.
     *
     * Source: CLIProxyAPI internal/auth/xai/{xai,types}.go and Management Center
     * src/utils/quota/constants.ts.
     */
    object Xai {
        const val CLIENT_ID = "b1a00492-073a-47ea-816f-4c329264a828"
        const val ISSUER = "https://auth.x.ai"

        /** Endpoints are discovered rather than hardcoded; see XaiProvider.discover(). */
        const val DISCOVERY_URL = "$ISSUER/.well-known/openid-configuration"

        const val SCOPE = "openid profile email offline_access grok-cli:access api:access"
        const val DEVICE_CODE_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code"

        const val BILLING_CREDITS_URL = "https://cli-chat-proxy.grok.com/v1/billing?format=credits"
        const val BILLING_URL = "https://cli-chat-proxy.grok.com/v1/billing"
        const val ME_URL = "https://api.x.ai/v1/me"

        const val CLIENT_VERSION = "0.2.91"
        const val USER_AGENT = "grok-pager/$CLIENT_VERSION grok-shell/$CLIENT_VERSION (android; aarch64)"

        val IDENTITY_HEADERS = mapOf(
            "x-xai-token-auth" to "xai-grok-cli",
            "x-grok-client-version" to CLIENT_VERSION,
        )
    }

    /**
     * Kimi Code.
     *
     * OAuth, under this app's OWN name. Kimi's RFC 8628 device flow uses one public client id
     * for every program that drives it — the first-party CLI, CLIProxyAPI, pi and others all
     * present the same one — and what `api.kimi.com/coding` gates on is the `X-Msh-Platform`
     * header, which names the calling program. Moonshot allowlists third-party programs on
     * request and forbids exactly one thing: presenting another program's identity. So this
     * app says who it is, never `kimi_cli`, and asks to be allowlisted under that name. Until
     * that is granted the coding API may answer `403 access_terminated`, and a key from the
     * user's own console stays available as the other way in. See docs/providers-kimi.md.
     *
     * Source: CLIProxyAPI internal/auth/kimi/kimi.go.
     */
    object Kimi {
        /** The public client id every Kimi Code device-flow client presents. */
        const val CLIENT_ID = "17e5f671-d194-4dfb-9706-5516cb48c098"
        const val DEVICE_CODE_URL = "https://auth.kimi.com/api/oauth/device_authorization"
        const val TOKEN_URL = "https://auth.kimi.com/api/oauth/token"
        const val DEVICE_CODE_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code"

        const val USAGE_ENDPOINT = "https://api.kimi.com/coding/v1/usages"
        const val CONSOLE_URL = "https://www.kimi.com/code"

        /** This app's own name, sent wherever Kimi asks which program is calling. */
        const val PLATFORM = "UsageLimits"

        /** Who is calling, truthfully. The version is the app's, so the two never disagree. */
        fun identityHeaders(version: String): Map<String, String> = mapOf(
            "User-Agent" to "$PLATFORM/$version (Android)",
            "X-Msh-Platform" to PLATFORM,
            "X-Msh-Version" to version,
        )
    }

    /**
     * Devin / Cognition.
     *
     * Devin's CLI login is a PKCE authorization-code flow with an ephemeral loopback
     * redirect.  The quota call is the same Connect-RPC method used by the official
     * integration; the JSON encoding is supported by the endpoint and keeps the mobile
     * client independent of a generated protobuf runtime.
     *
     * Source: CLIProxyAPI internal/auth/devin/{devin_auth,user_status}.go and the
     * CLIProxyAPI Management Center quota provider.
     */
    object Devin {
        const val APP_BASE_URL = "https://app.devin.ai"
        const val API_BASE_URL = "https://api.devin.ai"
        const val SERVER_BASE_URL = "https://server.codeium.com"
        const val AUTHORIZE_PATH = "/auth/cli/continue"
        const val TOKEN_PATH = "/auth/cli/token"
        const val PROFILE_PATH = "/v3/self"
        const val STATUS_PATH = "/exa.seat_management_pb.SeatManagementService/GetUserStatus"

        const val USER_AGENT = "UsageLimits/1.0"
        const val IDE_NAME = "chisel"
        const val IDE_VERSION = "3000.10.21"
        const val LOCALE = "en"

        const val SESSION_TOKEN_PREFIX = "devin-session-token\$"
    }

    /**
     * Meta Muse.
     *
     * Meta exposes OAuth as an RFC 8628 device flow.  The device token is retained separately
     * from the short-lived/minted LLM key: the same `muse-code/key` endpoint returns both the
     * subscription quota and a usable API key, but only the DCA token is valid for that query.
     *
     * Source: CLIProxyAPI internal/auth/meta/meta.go and the CLIProxyAPI Management Center
     * quota provider (Meta Muse).
     */
    object Meta {
        const val AUTH_HOST = "https://auth.meta.com"
        const val DEVICE_AUTHORIZATION_ENDPOINT = "$AUTH_HOST/oidc/device/authorization/"
        const val TOKEN_ENDPOINT = "$AUTH_HOST/oidc/device/token/"
        const val CLIENT_ID = "1031625952748946"
        const val DEVICE_CODE_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code"

        const val KEY_ENDPOINT = "https://api.meta.ai/muse-code/key"
        const val API_BASE_URL = "https://api.meta.ai/v1"
        const val USER_AGENT = "muse-code/1.0.2"
        const val API_VERSION = "1.0.0"
    }
}
